File: agent/api/poster_copy_sets.py
"""Poster Copy Set + AI Poster Copy Assistant API (POSTER_BUILDER_V2).

Poster-NATIVE copy domain endpoints — fully separate from the video
/api/copy-sets namespace. Nothing here reads or writes the video copy_set table.
"""
from __future__ import annotations

from typing import Any, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from agent.models.poster_copy_set import (
    PosterCopySetApproveRequest,
    PosterCopySetCreateRequest,
    PosterCopySetPatchRequest,
    PosterCopySetRejectRequest,
)
from agent.services import poster_copy_ai_service as ai_svc
from agent.services.poster_copy_set_service import (
    PosterCopySetError,
    PosterCopySetService,
)

router = APIRouter(prefix="/poster/copy-sets", tags=["poster-copy-sets"])


def _http(exc: PosterCopySetError | ai_svc.PosterCopyAIError) -> HTTPException:
    detail: dict[str, Any] = {"code": exc.code, "message": str(exc)}
    field_errors = getattr(exc, "field_errors", None)
    if field_errors:
        detail["field_errors"] = field_errors
    return HTTPException(status_code=exc.status_code, detail=detail)


class RecommendObjectivesRequest(BaseModel):
    product_id: str
    refresh_ai: bool = False


class RecommendAnglesRequest(BaseModel):
    product_id: str
    archetype: str
    refresh_ai: bool = False


class GenerateDirectionsRequest(BaseModel):
    product_id: str
    archetype: str
    angle: str
    tone: str = ""
    language: str = "ms"
    count: int = Field(default=3, ge=1, le=5)


class RegenerateFieldRequest(BaseModel):
    product_id: str
    archetype: str
    angle: str
    field: str
    language: str = "ms"
    fields: dict[str, Any] = Field(default_factory=dict)


@router.post("/recommend-objectives")
async def recommend_objectives(req: RecommendObjectivesRequest):
    try:
        return await ai_svc.recommend_objectives(req.product_id, refresh_ai=req.refresh_ai)
    except ai_svc.PosterCopyAIError as exc:
        raise _http(exc)


@router.post("/recommend-angles")
async def recommend_angles(req: RecommendAnglesRequest):
    try:
        return await ai_svc.recommend_angles(
            req.product_id, req.archetype, refresh_ai=req.refresh_ai
        )
    except ai_svc.PosterCopyAIError as exc:
        raise _http(exc)


@router.post("/directions")
async def generate_directions(req: GenerateDirectionsRequest):
    try:
        return await ai_svc.generate_directions(
            req.product_id,
            req.archetype,
            req.angle,
            tone=req.tone,
            language=req.language,
            count=req.count,
        )
    except ai_svc.PosterCopyAIError as exc:
        raise _http(exc)


@router.post("/regenerate-field")
async def regenerate_field(req: RegenerateFieldRequest):
    try:
        return await ai_svc.regenerate_field(
            req.product_id,
            req.archetype,
            req.angle,
            req.fields,
            req.field,
            language=req.language,
        )
    except ai_svc.PosterCopyAIError as exc:
        raise _http(exc)


@router.post("")
async def create_poster_copy_set(req: PosterCopySetCreateRequest):
    try:
        return await PosterCopySetService.create_draft(req)
    except PosterCopySetError as exc:
        raise _http(exc)


@router.get("")
async def list_poster_copy_sets(product_id: str):
    return {"poster_copy_sets": await PosterCopySetService.list_for_product(product_id)}


@router.get("/{poster_copy_set_id}")
async def get_poster_copy_set(poster_copy_set_id: str):
    try:
        return await PosterCopySetService.get(poster_copy_set_id)
    except PosterCopySetError as exc:
        raise _http(exc)


@router.patch("/{poster_copy_set_id}")
async def patch_poster_copy_set(
    poster_copy_set_id: str, req: PosterCopySetPatchRequest
):
    try:
        return await PosterCopySetService.patch_draft(poster_copy_set_id, req)
    except PosterCopySetError as exc:
        raise _http(exc)


@router.post("/{poster_copy_set_id}/new-version")
async def new_poster_copy_set_version(
    poster_copy_set_id: str, req: PosterCopySetPatchRequest
):
    """Safe edit flow for APPROVED sets: atomic child DRAFT + parent SUPERSEDED."""
    try:
        return await PosterCopySetService.new_version(poster_copy_set_id, req)
    except PosterCopySetError as exc:
        raise _http(exc)


@router.post("/{poster_copy_set_id}/fork-historical")
async def fork_poster_copy_set_from_historical(
    poster_copy_set_id: str, req: PosterCopySetPatchRequest
):
    """Reopen a saved poster whose copy set is now SUPERSEDED: clone the exact
    historical copy into a fresh DRAFT without mutating the historical record."""
    try:
        return await PosterCopySetService.fork_from_historical(poster_copy_set_id, req)
    except PosterCopySetError as exc:
        raise _http(exc)


@router.post("/{poster_copy_set_id}/approve")
async def approve_poster_copy_set(
    poster_copy_set_id: str, req: PosterCopySetApproveRequest
):
    try:
        return await PosterCopySetService.approve(
            poster_copy_set_id,
            approval_phrase=req.approval_phrase,
            approved_by=req.approved_by,
        )
    except PosterCopySetError as exc:
        raise _http(exc)


@router.post("/{poster_copy_set_id}/reject")
async def reject_poster_copy_set(
    poster_copy_set_id: str, req: PosterCopySetRejectRequest
):
    try:
        return await PosterCopySetService.reject(poster_copy_set_id, reason=req.reason)
    except PosterCopySetError as exc:
        raise _http(exc)

File: dashboard/src/api/posterCopySets.ts
// Poster Copy Set + AI Poster Copy Assistant + compositor API clients
// (POSTER_BUILDER_V2). Poster-native domain — separate from /api/copy-sets.
import { getAPI, patchAPI, postAPI } from "./client";
import type {
	PosterAngleRecommendation,
	PosterComposeResponse,
	PosterCopyDirection,
	PosterCopySet,
	PosterDeliverableReconstruction,
	PosterObjectiveRecommendation,
} from "../types/posterCopySet";

export async function recommendPosterObjectives(payload: {
	product_id: string;
	refresh_ai?: boolean;
}): Promise<{
	recommendations: PosterObjectiveRecommendation[];
	warnings: string[];
}> {
	return postAPI("/poster/copy-sets/recommend-objectives", payload);
}

export async function recommendPosterAngles(payload: {
	product_id: string;
	archetype: string;
	refresh_ai?: boolean;
}): Promise<{ angles: PosterAngleRecommendation[]; warnings: string[] }> {
	return postAPI("/poster/copy-sets/recommend-angles", payload);
}

export async function generatePosterDirections(payload: {
	product_id: string;
	archetype: string;
	angle: string;
	tone?: string;
	language?: string;
	count?: number;
}): Promise<{
	directions: PosterCopyDirection[];
	ai_model: string;
	prompt_version: string;
	warnings: string[];
}> {
	return postAPI("/poster/copy-sets/directions", payload);
}

export async function regeneratePosterField(payload: {
	product_id: string;
	archetype: string;
	angle: string;
	field: string;
	language?: string;
	fields: Record<string, unknown>;
}): Promise<{ field: string; value: string | string[]; provenance: string }> {
	return postAPI("/poster/copy-sets/regenerate-field", payload);
}

export async function createPosterCopySet(
	payload: Record<string, unknown>,
): Promise<PosterCopySet> {
	return postAPI("/poster/copy-sets", payload);
}

export async function approvePosterCopySet(
	posterCopySetId: string,
	approvalPhrase: string,
): Promise<PosterCopySet> {
	return postAPI(`/poster/copy-sets/${posterCopySetId}/approve`, {
		approval_phrase: approvalPhrase,
		approved_by: "operator",
	});
}

export async function listPosterCopySets(
	productId: string,
): Promise<{ poster_copy_sets: PosterCopySet[] }> {
	return getAPI(`/poster/copy-sets?product_id=${encodeURIComponent(productId)}`);
}

// ── Deterministic compositor (credit-free) ──────────────────────────────────

export async function composePoster(payload: {
	product_id: string;
	poster_copy_set_id: string;
	recipe_id: string;
	background_media_id?: string;
	settings?: Record<string, unknown>;
}): Promise<PosterComposeResponse> {
	return postAPI("/poster/compose", payload);
}

export async function savePosterToLibrary(
	posterDeliverableId: string,
): Promise<{ creative_asset_id: string; already_saved: boolean }> {
	return postAPI(`/poster/deliverables/${posterDeliverableId}/save-to-library`, {});
}

export function posterDeliverableOutputUrl(posterDeliverableId: string): string {
	return `/api/poster/deliverables/${posterDeliverableId}/output`;
}

// Creative Library round trip: reopen a saved poster from its asset id.
export async function fetchPosterDeliverableByAsset(
	creativeAssetId: string,
): Promise<PosterDeliverableReconstruction> {
	return getAPI(
		`/poster/deliverables/by-asset/${encodeURIComponent(creativeAssetId)}`,
	);
}

// Safe edit flow for an APPROVED set: creates DRAFT v(n+1), parent SUPERSEDED.
export async function newPosterCopySetVersion(
	posterCopySetId: string,
	patch: Record<string, unknown>,
): Promise<PosterCopySet> {
	return postAPI(`/poster/copy-sets/${posterCopySetId}/new-version`, patch);
}

export async function patchPosterCopySet(
	posterCopySetId: string,
	patch: Record<string, unknown>,
): Promise<PosterCopySet> {
	return patchAPI(`/poster/copy-sets/${posterCopySetId}`, patch);
}
